### PR TITLE
[SYCL][Joint Matrix][NFC] Add SYCLPropagateJointMatrixUsage pass

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
This patch adds a pass which propagates optional kernel features 
metadata through a module call graph for sycl_ext_oneapi_matrix 
extension. According to the extension spec, optional kernel 
features are the `joint_matrix` type and the `joint_matrix_mad` 
function.

### CODE REVIEW

File: llvm/lib/Passes/PassBuilder.cpp
Comment (dm-vodopyanov) [inline llvm/lib/Passes/PassBuilder.cpp:248]: Weird. clang-format requirement.

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:186]:
```suggestion
void setSyclJointMatrixMetadata(StringRef MetadataName, Module *M,
```

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:190]:
```suggestion
  SmallString<256> StringValue;
```

[Doxygen](https://llvm.org/doxygen/classllvm_1_1SmallString.html)
No need to re-allocate the whole string on each `+=`

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:198]: Could it be done in form of an early exit `if (Values.empty()) return;` ?

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:118]:
```suggestion
  if (!F->hasFnAttribute("sycl-joint-matrix-mad-type-A"))
    return;
  // less indentation for the rest of the code
  // early exits is our everything
```

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:119]: [`llvm::SmallString`](https://llvm.org/doxygen/classllvm_1_1SmallString.html)

`+=` is expensive for `std::string`

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:84]: See comments on `fillFunctionToJointMatrixMapValuesMap`

File: llvm/lib/Passes/PassBuilder.cpp
Comment (dm-vodopyanov) [reply llvm/lib/Passes/PassBuilder.cpp:249]: Right, 
```
#include "llvm/Transforms/Utils/DXILUpgrade.h"
```
line was added in llorg, so clang-format check was avoided. I included `SYCLPropagateJointMatrixUsage.h` above and that's why clang-format triggered on this. This incorrect placing of the header will bring us one more merge conflict during the internal pulldown.

**Comment by dm-vodopyanov** [comment]
@dkhaldi, @YuriPlyakhin, @yubingex007-a11y, @intel/dpcpp-tools-reviewers, @intel/dpcpp-esimd-reviewers, @intel/llvm-reviewers-runtime, @intel/dpcpp-cfe-reviewers, @aelovikov-intel , this is ready to be reviewed.

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (YuriPlyakhin) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:115]: Did you mean to use mad in the name?
```suggestion
void fillFunctionToJointMatrixMadValuesMap(
```

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (YuriPlyakhin) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:117]:
```suggestion
    FunctionToJointMatrixValuesMapTy &FunctionToJointMatrixMadValues) {
```

**Comment by YuriPlyakhin** [comment]
Question for my understanding: why does this pass cover only c'tor and mad, but no other matrix api calls, like load, store, fill, etc...

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (dm-vodopyanov) [reply llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp]: Yeah, good catch!

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (dkhaldi) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:102]: should we add scope as well here? so if a matrix is declared as 
joint_matrix<work_group, half, use::a, 8,16, layout::row_major> A;
this is not supported as only sub_group Group is supported.

File: sycl/test/matrix/matrix-check-types-in-attributes.cpp
Comment (dkhaldi) [inline sycl/test/matrix/matrix-check-types-in-attributes.cpp:5]: please add a comment about what 12, 12 mean here.

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (dkhaldi) [reply llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:102]: I agree, I will track adding this in the spec. We removed scope from the spec as we only have one value supported.

Comment (AlexeySachkov) [reply llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:102]: That specification is not for error checking, it is for understanding which features are used by a certain kernel. 

Checking that `work_group` is a wrong value to pass into the first template arguments could be done with a simple `static_assert` in headers, no need to bring this into the compiler.

File: sycl/test/matrix/matrix-check-types-in-attributes.cpp
Comment (aelovikov-intel) [inline sycl/test/matrix/matrix-check-types-in-attributes.cpp:1]: Please fix CODEOWNERS in a separate PR to add "sycl/test/matrix" to matrix reviewers ownership. Once merged in a separate PR, this one shouldn't require SYCL RT approval anymore (I think).

Comment (dm-vodopyanov) [reply sycl/test/matrix/matrix-check-types-in-attributes.cpp:1]: Created https://github.com/intel/llvm/pull/11564

**Comment by dm-vodopyanov** [comment]
@intel/dpcpp-esimd-reviewers @intel/dpcpp-cfe-reviewers - friendly ping

File: sycl/include/sycl/ext/oneapi/matrix/query-types.hpp
Comment (yubingex007-a11y) [inline sycl/include/sycl/ext/oneapi/matrix/query-types.hpp:50]: could we use traits to make the code cleaner?
```
template <typename T>
struct traits {
    static constexpr const char* value = "";
};

#define traits(T, String)                                    \
  template <> struct traits<T> {                              \
    static constexpr const char* value = String;                         \
  };

traits(int, "matrix_type::fp32")
```
https://godbolt.org/z/5q5MGdM9d

Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/matrix/query-types.hpp:50]: I don't see how "struct" is better than function here. I'd make a single function and use "if constexpr (std::is_same_v<T, Type1>) {}" though.

Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/query-types.hpp:50]: @aelovikov-intel looks good!

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (yubingex007-a11y) [inline clang/lib/CodeGen/BackendUtil.cpp:973]: i have a question here. usually when we add a pass, we need to modify a file such as llvm/test/CodeGen/X86/O0-pipeline.ll. why we don't have such check here?

File: sycl/include/sycl/ext/oneapi/matrix/query-types.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/ext/oneapi/matrix/query-types.hpp:50]: > I don't see how "struct" is better than function here. I'd make a single function and use "if constexpr (std::is_same_v<T, Type1>) {}" though.

I did it as a single function initially in the early commits but didn't like these else if/else if/else if, so changed to the template specializations.

Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/matrix/query-types.hpp:50]: I don't agree - https://godbolt.org/z/MYrdE64n7. That said, this is up to the "matrix" reviewers.

**Review by sarnex — APPROVED** [review]
Reviewed `llvm/lib/SYCLLowerIR/CMakeLists.txt` as esimd code owner, did not review the rest, that file LGTM

File: sycl/include/sycl/ext/oneapi/matrix/query-types.hpp
Comment (dm-vodopyanov) [reply sycl/include/sycl/ext/oneapi/matrix/query-types.hpp:50]: I meant the pure code. With the macro it looks very clean, of course.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (elizabethandrews) [inline clang/lib/CodeGen/BackendUtil.cpp:969]: Why the additional braces? Formatting rules prefer consistency here. Either no braces for the `if `or we need braces for all the branches i.e. `else `as well.

Comment (dm-vodopyanov) [reply clang/lib/CodeGen/BackendUtil.cpp:969]: They are not needed, the legacy after a try to add a pass just after the if statement, will be removed.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (AlexeySachkov) [reply clang/lib/CodeGen/BackendUtil.cpp:973]: The pass is only added when we compiling for SYCL device. I assume that the test you have mentioned checks some other target and therefore it doesn't trigger the new pass at all.

**Review by smanna12 — APPROVED** [review]

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (jzc) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:60]: I don't think so - since it's `sycl-post-link`'s choice of whether or these as entry points or not, then even if this information is propagated, it can be discarded by sycl-post-link. By the way, can a test be added to ensure the joint matrix features are propagated?

Comment (dm-vodopyanov) [reply llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:60]: Sounds reasonable.

> By the way, can a test be added to ensure the joint matrix features are propagated to SYCL_EXTERNAL functions?

The test already covers SYCL_EXTERNAL functions.

Comment (jzc) [reply llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:60]: Oh, I actually didn't see that some of the functions in `call-graph-joint-matrix.ll` had `sycl-module-id`. Even though, the functions that do have `sycl-module-id` would have had joint matrix attributes propagated to them anyways because they are called from `spir_kernel` functions. The case that I had in mind and that was not covered in `call-graph-joint-matrix.ll` is when a `spir_func` with `sycl-module-id` that is not called from a `spir_kernel` has joint matrix features.

Comment (dm-vodopyanov) [reply llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:60]: Thanks, added such case

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:197]:
```suggestion
  for (auto It = Values.begin(); It != Values.end(); It++) {
```

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:218]:
```suggestion
  SmallVector<Function *, 16> EntryPoints;
```

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:2]:
```suggestion
//===---- SYCLPropagateJointMatrixUsage.cpp -------------------------------===//
```

I would maybe suggest to drop this small "description" completely since it doesn't fit.
You can put it on the next line, but considering that it doesn't bring a lot of extra information, I think that we can just drop it.

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:93]: We should probably have a `using` for this `SmallString<24>` to avoid repeating magic number `24`. 

Also, why 24?

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp:48]: We probably could have just done a bottom-up call graph scan by going through function uses to avoid copying all this code here and reason about what is entry point and what is not. However, I'm fine with the current approach as well.

If that would be necessary, we can refactor later once we get a bit more experience with this functionality.

**Review by AlexeySachkov — APPROVED** [review]

File: llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp
Comment (dm-vodopyanov) [reply llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp]: Actually, it is better to have like 40, 24 is outdated. This number is the average size of the string which will be generated from matrix type strings and matrix size strings

**Pull request merged**
